### PR TITLE
Low level codegen

### PR DESCRIPTION
**Pull request by jianghaolu**

### CODE REVIEW

**Comment by jianghaolu** [comment]
Need to support serviceVersion per comment https://github.com/Azure/azure-sdk-for-java/pull/20866#discussion_r618596076. But will do in a separate PR

**Comment by weidongxu-microsoft** [comment]
Any sample code on the generated low level client? I am not familiar with the concept yet.

**Comment by jianghaolu** [comment]
> Any sample code on the generated low level client? I am not familiar with the concept yet.
> 
> Ok, forget it. https://github.com/Azure/azure-sdk-for-java/pull/20866/files would be a sample.

Yes that would be a good one

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java [hunk 1]
```diff
@@ -292,6 +295,21 @@ private ClientResponse parseResponse(Operation method, JavaSettings settings) {
...
+        // WARNING: Only tested for low level clients
+        JavaSettings settings = JavaSettings.getInstance();
+        ModuleInfo moduleInfo = new ModuleInfo(settings.getPackage());
+
+        List<ModuleInfo.RequireModule> requireModules = moduleInfo.getRequireModules();
+        requireModules.add(new ModuleInfo.RequireModule("com.azure.core", true));
+        requireModules.add(new ModuleInfo.RequireModule("com.azure.core.experimental", true));
+
+        List<ModuleInfo.ExportModule> exportModules = moduleInfo.getExportModules();
+        exportModules.add(new ModuleInfo.ExportModule(settings.getPackage()));
```
Comment (srnagar) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java:308]: If the LLC client is provided with a package name that has `implementation` in it, we should not export it. Should we even support generating LLC if the package name has `implementation`?

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java [hunk 2]
```diff
@@ -292,6 +295,21 @@ private ClientResponse parseResponse(Operation method, JavaSettings settings) {
         return builder.build();
     }
 
+    private static ModuleInfo moduleInfo() {
```
Comment (srnagar) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java:298]: Do we generate `pom.xml` too for LLC?

**Review by srnagar — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/mapper/ServiceClientMapper.java [hunk 3]
```diff
@@ -139,10 +139,10 @@ public ServiceClient map(CodeModel codeModel) {
             }
         }
         serviceClientProperties.add(new ServiceClientProperty("The HTTP pipeline to send requests through.",
-                ClassType.HttpPipeline, "pipeline", true, null));
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ServiceClientMapper.java]: If you do intend to use "pipeline", I can accommodate it in fluentgen, as it is not exposed to customer so not counted as breaking changes (though better not to change if no very good reason).

Comment (jianghaolu) [reply javagen/src/main/java/com/azure/autorest/mapper/ServiceClientMapper.java]: it's the inconsistency between the naming in the client and the builder that drove me to make the change - but feel like it's probably less painful to just avoid this change

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java (hunk 2 above)
Comment (jianghaolu) [reply javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java:298]: no we don't

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java (hunk 1 above)
Comment (jianghaolu) [reply javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java:308]: If --low-level-client is turned on, the only 3 files we generate are the client, the builder, and the module-info. I don't quite understand why someone would put the client and builder in `implementation`?

**Comment by jianghaolu** [comment]
Pending `RequestOptions` being added to `azure-core-experimental`

**Comment by jianghaolu** [comment]
Work is moved to the `llc` branch.

**Pull request closed without merging**